demos/enclave/ETL/compile_models.py

```diff
@@ -0,0 +1,411 @@
+from datetime import datetime
+import json
+from typing import Literal
+from uuid import UUID, uuid4
+import os
+import boto3
+import argparse
+import pandas as pd
+from io import BytesIO
+from zipfile import ZipFile
+from pydantic import BaseModel, Extra, validator
+
+MODEL_FILE_USERS = "users.csv"
+MODEL_FILE_COURSES = "courses.csv"
+MODEL_FILE_ONEROSTER_DEMOGRAPHICS = "oneroster_demographics.csv"
+MODEL_FILE_ENROLLMENTS = "enrolments.csv"
+MODEL_FILE_ASSESSMENTS = "assessments.csv"
+MODEL_FILE_GRADES = "grades.csv"
+
+
+class Demographic(BaseModel):
+    user_uuid: UUID
+    birth_date: str
+    sex: Literal['male', 'female']
+    american_indian_or_alaska_native: Literal['true', 'false']
+    asian: Literal['true', 'false']
+    black_or_african_american: Literal['true', 'false']
+    native_hawaiian_or_other_pacific_islander: Literal['true', 'false']
+    white: Literal['true', 'false']
+    demographic_race_two_or_more_races: Literal['true', 'false']
+    hispanic_or_latino_ethnicity: Literal['true', 'false']
+
+    class Config:
+        extra = Extra.forbid
+
+    @validator('birth_date')
+    def birthdate_format(cls, v):
+        datetime.strptime(v, "%Y-%m-%d")
+        return v
+
+
+class Assessment(BaseModel):
+    id: int
+    name: str
+
+    class Config:
+        extra = Extra.forbid
+
+
+class Course(BaseModel):
+    id: int
+    name: str
+
+    class Config:
+        extra = Extra.forbid
+
+
+class Enrollment(BaseModel):
+    user_uuid: UUID
+    course_id: int
+    role: Literal['student', 'teacher']
+
+    class Config:
+        extra = Extra.forbid
+
+
+class Grades(BaseModel):
+    assessment_id: int
+    user_uuid: UUID
+    course_id: int
+    grade_percentage: float
+
+    class Config:
+        extra = Extra.forbid
+
+    @validator('grade_percentage')
+    def grade_value(cls, v):
```

Comment:
I was wondering how the validator didn't catch the fact that some grades have `grade_percentage` as an empty value. It seems that the Pydantic validator accepts `numpy.nan` as a valid value per [here](https://github.com/pydantic/pydantic/issues/1779) and serializes that as an empty string to the CSV later. We can modify this validator a bit to catch these, though:

```python
    @validator('grade_percentage')
    def grade_value(cls, v):
        if isnan(v):
            raise ValueError(f'Grade value is nan')
        if v < 0.0 or v > 100.0:
            raise ValueError(f'Grade value {v} is out of expected range')
        return v
```

The `isnan` can be imported from `math`:

```python
from math import isnan
```